### PR TITLE
feat: Configured extra check for waitlisted users already in the list and created toast message for them

### LINKED ISSUE #463: WEB: suggesting changes while adding waitlist
# Description 
- [x] [chore] Right now after adding the user to the waitlist there the text field remains the same, instead lets clear the text field. (#481)
- [x] [chore] Perform some validation on the email field as `ff@ff` and other things is allowed which should not be allowed. (#487)
- [x] [chore] Only show toast when the whitelist response is successful. (#490)
- [ ] [feat] Check if user has already whitelisted. (#492)

---
Target file: https://github.com/keyshade-xyz/keyshade/blob/72281e66569b05011a7b79a94ae502eaac0b3a6d/apps/web/src/components/hero/index.tsx

### PR DESCRIPTION
**Pull request by poswalsameer**
## Description
This PR solves the issue by showing an error/warning toast message to already existing users on the waitlist. Before this PR, every user with a single email was getting the success toast message multiple times. Ideally, this should not happen, every user should be able to join the waitlist one time through one email. This PR fixes this problem.

Fixes #463 

## Dependencies
No dependencies or packages are used.

## Future Improvements
Right now, the data for waitlisted users is stored in the local storage but this logic can be improved by mapping them to a database in the future.

## Screenshots of relevant screens
Screenshot (94)

## Developer's checklist

- [ x] My PR follows the style guidelines of this project
- [ x] I have performed a self-check on my work

**If changes are made in the code:**

- [ x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [ x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ x] I have added relevant screenshots in my PR
- [ x] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ x] I have made the necessary updates to the documentation, or no documentation changes are required.

___

### **PR Type**
enhancement, bug_fix

___

### **Description**
- Implemented a check to verify if a user's email is already in the waitlist, preventing duplicate entries.
- Displayed a custom toast message to inform users if they are already on the waitlist.
- Updated the local storage to include new emails added to the waitlist.
- Cleared the email input field after a successful waitlist addition to improve user experience.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>index.tsx</strong><dd><code>Add email validation and waitlist check with toast notifications</code></dd></summary>
<hr>

apps/web/src/components/hero/index.tsx

<li>Added logic to check if an email is already in the waitlist.<br> <li> Displayed a custom toast message for already waitlisted users.<br> <li> Updated local storage with new waitlisted emails.<br> <li> Cleared the email input field after successful submission.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/492/files#diff-7613ce5c9e898c20a2427ba5bf13810f61bb0a532139896e00a54cd0285ad40c">+26/-2</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by poswalsameer** [comment]
Hi @rajdip-b and @kriptonian1, this is a fresh PR that I have created for issue #463. You can review the code and let me know if any more changes are required. Thanks, loved working on this!

**Comment by rajdip-b** [comment]
Can you resolve the conflict?

**Comment by poswalsameer** [comment]
> Can you resolve the conflict?

Let me see.

**Comment by poswalsameer** [comment]
> Can you resolve the conflict?

Cannot figure out why is this showing as a conflict here, because that is the code which I have added to add the email in the state ( inside the input box ). Would like a small help in this!

**Comment by kriptonian1** [comment]
> > Can you resolve the conflict?
> 
> Cannot figure out why is this showing as a conflict here, because that is the code which I have added to add the email in the state ( inside the input box ). Would like a small help in this!

Would you like to get on a call on discord?

**Comment by poswalsameer** [comment]
> @poswalsameer Would you like to get on a call on discord?

yeah sure

**Comment by poswalsameer** [comment]
Hi @kriptonian1, done with the changes as we discussed. You can have a final look.

File: apps/web/src/components/hero/index.tsx [hunk 1]
```diff
@@ -44,8 +60,13 @@ function Hero(): React.JSX.Element {
               launch
             </p>
           </div>
-        ))
+
+        ))    
+
+        waitListedEmails.push(email)
+        localStorage.setItem('waitListData', JSON.stringify(waitListedEmails))
```
Comment (rajdip-b) [inline apps/web/src/components/hero/index.tsx:67]: Rename `waitListData` to `waitlistData`

File: apps/web/src/components/hero/index.tsx [hunk 2]
```diff
@@ -27,6 +27,22 @@ function Hero(): React.JSX.Element {
       return
     }
 
+    const dataInStorage: string | null = localStorage.getItem('waitListData')
+    const waitListedEmails: string[] = dataInStorage ? (JSON.parse(dataInStorage) as string[]) : []
```
Comment (rajdip-b) [inline apps/web/src/components/hero/index.tsx:31]: Rename `waitListedEmails` to `emailsInWaitlist`

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: apps/web/src/components/hero/index.tsx [hunk 3]
```diff
@@ -27,6 +27,22 @@ function Hero(): React.JSX.Element {
...
+  
+    // actual logic where we are checking if this email is already in waitlisted users or not
+    if (waitListedEmails.includes(email)) {
+      toast.custom(() => (
+        <div className="text-brandBlue border-brandBlue/20 w-[90vw] rounded-lg border bg-[#852b2c] p-2 shadow-2xl backdrop-blur-3xl md:w-[20vw]">
+          <p className="text-sm">
+            You have been already added to the waitlist. We will notify you once
+            we launch.{' '}
```
Comment (rajdip-b) [inline apps/web/src/components/hero/index.tsx:39]: Not sure why we would need the `{' '}`

File: apps/web/src/components/hero/index.tsx [hunk 4]
```diff
@@ -27,6 +27,22 @@ function Hero(): React.JSX.Element {
       return
     }
 
+    const dataInStorage: string | null = localStorage.getItem('waitListData')
+    const waitListedEmails: string[] = dataInStorage ? (JSON.parse(dataInStorage) as string[]) : []
+  
+    // actual logic where we are checking if this email is already in waitlisted users or not
+    if (waitListedEmails.includes(email)) {
+      toast.custom(() => (
+        <div className="text-brandBlue border-brandBlue/20 w-[90vw] rounded-lg border bg-[#852b2c] p-2 shadow-2xl backdrop-blur-3xl md:w-[20vw]">
```
Comment (rajdip-b) [inline apps/web/src/components/hero/index.tsx:36]: Not an expert but do we not have any theme color for `bg-[#852b2c]`

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: apps/web/src/components/hero/index.tsx (hunk 4 above)
Comment (poswalsameer) [reply apps/web/src/components/hero/index.tsx:36]: I have used this color from the toast message which is already been used to show errors. That's why went with this bg. Rest of the changes in the variable names have been done successfully and pushed to the same branch.

Comment (kriptonian1) [reply apps/web/src/components/hero/index.tsx:36]: @poswalsameer will it be possible for you or add this hex code to the tailwind config so we can have uniformity

**Review by kriptonian1 — APPROVED** [review]
LGTM, but we will really appreciate it if you can do the last request we made

File: apps/web/src/components/hero/index.tsx (hunk 4 above)
Comment (poswalsameer) [reply apps/web/src/components/hero/index.tsx:36]: > @poswalsameer will it be possible for you or add this hex code to the tailwind config so we can have uniformity

Sure, that won't be an issue for me. What name should I go with? I was thinking of naming it "errorRed", please let me know about this.

Comment (kriptonian1) [reply apps/web/src/components/hero/index.tsx:36]: You can name it anything relevant

Comment (poswalsameer) [reply apps/web/src/components/hero/index.tsx]: > You can name it anything relevant

Hi, done with the change, added a variable for errorRed in tailwind.config.js. You can see the latest commit.

**Comment by kriptonian1** [comment]
@rajdip-b if everything is okey from your end then we can merge

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**